planner/core/rule_partition_processor.go

```diff
@@ -694,8 +696,110 @@ func pruneUseBinarySearch(lessThan lessThanDataInt, data dataForPrune, unsigned
 	return start, end
 }
 
+func (s *partitionProcessor) resolveAccessPaths(ds *DataSource) error {
+	possiblePaths, err := getPossibleAccessPaths(
+		ds.ctx, &tableHintInfo{indexMergeHintList: ds.indexMergeHints, indexHintList: ds.IndexHints},
+		ds.astIndexHints, ds.table, ds.DBName, ds.tableInfo.Name)
+	if err != nil {
+		return err
+	}
+	possiblePaths, err = filterPathByIsolationRead(ds.ctx, possiblePaths, ds.DBName)
+	if err != nil {
+		return err
+	}
+	ds.possibleAccessPaths = possiblePaths
+	return nil
+}
+
+func (s *partitionProcessor) resolveOptimizeHint(ds *DataSource, partitionName model.CIStr) error {
+	// index hint
+	if len(ds.IndexHints) > 0 {
+		newIndexHint := make([]indexHintInfo, 0, len(ds.IndexHints))
+		for _, idxHint := range ds.IndexHints {
+			if len(idxHint.partitions) == 0 {
+				newIndexHint = append(newIndexHint, idxHint)
+			} else {
+				for _, p := range idxHint.partitions {
+					if p.String() == partitionName.String() {
+						newIndexHint = append(newIndexHint, idxHint)
+						break
+					}
+				}
+			}
+		}
+		ds.IndexHints = newIndexHint
+	}
+
+	// index merge hint
+	if len(ds.indexMergeHints) > 0 {
+		newIndexMergeHint := make([]indexHintInfo, 0, len(ds.indexMergeHints))
+		for _, idxHint := range ds.indexMergeHints {
+			if len(idxHint.partitions) == 0 {
+				newIndexMergeHint = append(newIndexMergeHint, idxHint)
+			} else {
+				for _, p := range idxHint.partitions {
+					if p.String() == partitionName.String() {
+						newIndexMergeHint = append(newIndexMergeHint, idxHint)
+						break
+					}
+				}
+			}
+		}
+		ds.indexMergeHints = newIndexMergeHint
+	}
+
+	// read from storage hint
+	if ds.preferStoreType&preferTiKV > 0 {
+		if len(ds.preferPartitions[preferTiKV]) > 0 {
+			ds.preferStoreType ^= preferTiKV
+			for _, p := range ds.preferPartitions[preferTiKV] {
+				if p.String() == partitionName.String() {
+					ds.preferStoreType |= preferTiKV
+				}
+			}
+		}
+	}
+	if ds.preferStoreType&preferTiFlash > 0 {
+		if len(ds.preferPartitions[preferTiFlash]) > 0 {
+			ds.preferStoreType ^= preferTiFlash
+			for _, p := range ds.preferPartitions[preferTiFlash] {
+				if p.String() == partitionName.String() {
+					ds.preferStoreType |= preferTiFlash
+				}
+			}
+		}
+	}
+	if ds.preferStoreType&preferTiFlash != 0 && ds.preferStoreType&preferTiKV != 0 {
+		ds.ctx.GetSessionVars().StmtCtx.AppendWarning(
+			errors.New("hint `read_from_storage` has conflict storage type for the partition " + partitionName.L))
+	}
+
+	return s.resolveAccessPaths(ds)
+}
+
+func checkTableHintsApplicableForPartition(partitions []model.CIStr, ds *DataSource, partitionSet set.StringSet) {
+	for _, p := range partitions {
+		if !partitionSet.Exist(p.L) {
+			ds.ctx.GetSessionVars().StmtCtx.AppendWarning(
+				errors.New(fmt.Sprintf("partition `%s` in hint can't hit any source", p.L)))
```

Comment:
Can we return the partition name back to `checkHintsApplicable`, and build a more comprehensible error message like `Unknow partition %s in optimizer hint %s`